Add fail_on_nonzero_exit parameter to SSM operators for exit code routing

ksharlandjiev opened the pull request:
## Problem

SSM operators currently fail when commands return non-zero exit codes, making it impossible to:
- Route workflows based on different exit codes
- Handle commands where non-zero exit codes represent valid business states (e.g., partial success, warnings)
- Implement conditional retry logic based on specific exit codes
- Migrate from traditional schedulers like Autosys that support exit code routing

Users have been forced to implement manual polling workarounds with custom Python tasks to handle these scenarios.

## Proposal

Add a `fail_on_nonzero_exit` parameter (default: `True`) to `SsmRunCommandOperator`, `SsmRunCommandCompletedSensor`, and `SsmRunCommandTrigger`.

When set to `False`:
- Tasks complete successfully regardless of command exit codes
- Exit codes can be retrieved with `SsmGetCommandInvocationOperator` for routing decisions
- AWS-level failures (TimedOut, Cancelled) still raise exceptions
- Command-level failures (non-zero exit codes) are tolerated

The default value of `True` maintains existing behavior for backward compatibility.

Commit:
Add fail_on_nonzero_exit parameter to SsmRunCommandOperator

Add a new fail_on_nonzero_exit parameter to SsmRunCommandOperator
and SsmRunCommandTrigger to allow workflows to continue when SSM
commands return non-zero exit codes, enabling exit-code-based
workflow routing.

When fail_on_nonzero_exit=False:
- Command-level failures (non-zero exit codes) are tolerated
- AWS-level failures (Cancelled, TimedOut) still raise exceptions
- Helpful log messages indicate the command status and exit code

The parameter defaults to True to maintain backward compatibility
with existing DAGs. This change supports both traditional (sync)
and deferrable (async) execution modes.

Commit:
Add fail_on_nonzero_exit parameter to SSM sensor and trigger

Enhance SSM sensor and trigger to support exit code routing patterns
by adding a fail_on_nonzero_exit parameter that allows DAGs to handle
command-level failures gracefully while still failing on AWS-level errors.

Changes:
- Add fail_on_nonzero_exit parameter to SsmSensor with default True
- Add fail_on_nonzero_exit parameter to SsmRunCommandTrigger with default True
- Add parameter to trigger's serialized_fields for proper serialization
- Implement inline status checks to distinguish AWS-level vs command-level failures
- AWS-level failures (Cancelled, TimedOut) always raise exceptions
- Command-level failures (Failed with exit code) are tolerated when fail_on_nonzero_exit=False
- Add comprehensive logging for both error and success scenarios in enhanced mode
- Maintain 100% backward compatibility (default behavior unchanged)

Commit:
Document SsmGetCommandInvocationOperator for exit code routing

Add  documentation and tests to clarify that
SsmGetCommandInvocationOperator provides all functionality
needed for exit code-based workflow routing.

Changes:
- Enhanced operator docstring with exit code routing use case examples
- Added workflow pattern documentation showing integration with enhanced mode
- Documented return value structure highlighting response_code field
- Added test_exit_code_routing_use_case to demonstrate custom exit code retrieval
- Added test_multiple_exit_codes_for_routing for multi-instance scenarios

Commit:
Add  documentation for SSM exit code handling

Add detailed documentation for the enhanced exit code handling feature
in SSM operators and sensors, including usage patterns, migration guide,
and best practices.

The new documentation covers:
- Overview of enhanced vs traditional operational modes
- Four usage patterns with code examples (async, sync, routing, traditional)
- Complete parameter reference with behavior tables
- Migration guide from manual polling anti-patterns
- Best practices for different use cases
- Common use cases and troubleshooting guidance

Commit:
Add unit tests for SSM exit code handling enhancements

Commit:
ruff fix

Commit:
Add documentation to provider.yaml ; Fix spelling mistakes.

vincbeck reviewed (commented):
My general comments:
- I think the idea is great, that is indeed a feature user might want. Thanks for creating a PR for that
- Creating documentation is great, however, I think the document is way too long. This is only my personal opinion so I would wait to see what others think but if for new parameter we are creating documentation that big, this will be impossible to maintain. Using AI to create code and/or documentation is great but we should also keep in mind, the longer is NOT the better. Again, I support the documentation, but this is way too big to me. As a user I wont probably read it all, and as a developer I am scared we need to maintain that
- The system test is a great idea, could you please move these 3 examples in the current system test?

Commit:
Consolidate SSM exit code documentation into main SSM doc

Commit:
Consolidate SSM exit code tests into main system test

Commit:
Remove ssm_exit_codes.rst reference from provider.yaml

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/operators/ssm.py:
```diff
@@ -132,14 +143,36 @@ def execute(self, context: Context):
+                        status = invocation.get("Status", "")
+
+                        # AWS-level failures should always raise
+                        if status in ("Cancelled", "TimedOut"):
```
Are these the only "aws-level" failures? Always cautious of static lists like this.

ksharlandjiev replied:
Yes. I have double checked the documentation. However, to make things easier, I've added `SsmHook.is_aws_level_failure()` helper method with the statuses list.

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/operators/ssm.py:
```diff
@@ -132,14 +143,36 @@ def execute(self, context: Context):
+
+                        # AWS-level failures should always raise
+                        if status in ("Cancelled", "TimedOut"):
+                            self.log.error("Command failed with AWS-level error: %s", status)
```
We probably don't need this? It fails today without a log when these situations happen.

ksharlandjiev replied:
removed.

o-nikolas commented on providers/amazon/src/airflow/providers/amazon/aws/triggers/ssm.py:
```diff
@@ -89,14 +94,47 @@ async def run(self) -> AsyncIterator[TriggerEvent]:
+                        self.status_message,
+                        self.status_queries,
+                    )
+                except Exception:
```
It looks like you do this pattern three separate times with lots of duplicated log strings, state sets/tuples, etc. Can this logic be extracted out into a helper method that's defined once and used three times?

ksharlandjiev replied:
done. I've removed the code duplication across operator, sensor, and trigger

o-nikolas reviewed (commented):
I've just looked through the docs and source code so far (not tests yet). I like the shortened docs.

Just one question: Can this be done today with trigger rules?

Commit:
reducing volume of docs and structure in docs

Commit:
fix empty line with trailing whitespace.

ksharlandjiev committed:
Merge branch 'main' into ssm-exit-code-handling

ksharlandjiev committed:
Merge branch 'apache:main' into ssm-exit-code-handling

Commit:
refactor(aws/ssm): Extract AWS-level failure detection into reusable hook method

- Add `is_aws_level_failure()` static method to SsmHook to centralize AWS-level failure detection logic
- Update SsmRunCommandOperator to use the new hook method instead of inline status checks
- Update SsmRunCommandCompletedSensor to use the new hook method for consistency
- Update SsmRunCommandTrigger to use the new hook method for consistency
- Add comprehensive unit tests for `is_aws_level_failure()` covering all SSM command statuses
- Improve code maintainability by eliminating duplicate failure detection logic across operators, sensors, and triggers
- Clarify distinction between AWS-level failures (Cancelled, TimedOut, Cancelling) and command-level failures (Failed)

ksharlandjiev committed:
Merge branch 'main' into ssm-exit-code-handling

Commit:
fix MyPy checks

Commit:
Merge remote-tracking branch 'refs/remotes/origin/ssm-exit-code-handling' into ssm-exit-code-handling

Commit:
Added "Cancelling" to the spelling wordlist. "Cancelling" is official AWS SSM status value as per AWS documentation.

o-nikolas reviewed (approved):
Have you run the system test to ensue that it's working correctly?

ksharlandjiev committed:
Merge branch 'main' into ssm-exit-code-handling

vincbeck commented on providers/amazon/docs/operators/ssm.rst:
```diff
@@ -52,12 +52,42 @@ Waiter. Additionally, you can use the following components to track the status o
     :start-after: [START howto_operator_run_command]
     :end-before: [END howto_operator_run_command]
+
+Exit Code Handling
```
nit: I am not a big fan of capitalizing everything

```suggestion
Exit code handling
```

vincbeck reviewed (commented):
One nit but overall looks good!

ksharlandjiev committed:
Update providers/amazon/docs/operators/ssm.rst

Co-authored-by: Vincent <97131062+vincbeck@users.noreply.github.com>

Commit:
ruff fix
